internal/services/storage/storage_account_resource.go

```diff
@@ -1503,7 +1503,11 @@ func resourceStorageAccountUpdate(d *pluginsdk.ResourceData, meta interface{}) e
 		opts.AccountPropertiesUpdateParameters.DefaultToOAuthAuthentication = &defaultToOAuthAuthentication
 	}
 
-	if d.HasChange("cross_tenant_replication_enabled") {
+	// During the update from v2 to v3 of the provider, users are able to set this newly introduced property to be `false` (was absent in v2 provider).
+	// The storage service differentiate this property being `false` or absent, while the d.HasChange regards both (`false` and `null`) are the same.
+	// This makes setting the property to be `false` ends up being absent at the remote end.
+	// Hence, we need to explicitly check for the null case here to make it update the property to `false` if user has set so (instead of skipping the if branch below).
+	if d.HasChange("cross_tenant_replication_enabled") || d.GetRawState().GetAttr("cross_tenant_replication_enabled").IsNull() {
```

Comment:
Rather than doing this, it'd be better to add a state migration to ensure that this value always has a default value if it's not explicitly set?

FWIW we're intentionally not using `d.GetRawState()` directly, in favour of a wrapper in `pluginsdk.IsExplicitlyNullInConfig` - so that we can ensure this is behaviourally consistent for the Framework migration (for the Framework shim, we'll be exposing an interface similar [to this](https://github.com/hashicorp/terraform-provider-azurerm/blob/05b4df660ea519e0fd4e264e25ba20a2df2088f5/internal/sdk/resource_data.go#L3-L39) [which'll probably get an `IsExplicitlyNull` method and serving a Plugin SDK or Framework implementation depending on which we're serving under - so we're intentionally not exposing `GetRawState` directly to not need to transform these objects)

Reply:
@tombuildsstuff Thank you for the hint! In my case, it was `d.GetRawState()`, shall I create a similar warpper like `pluginsdk.IsNullInState`?

I've been skimmed on the implementation of `pluginsdk.IsExplicitlyNullInConfig`, which seems odd in case the field is not defined in the config it returns `true`, which makes a non-exist field (in terms of schema) also returns `true` in this case. Instead, shall we do something like:

```go
func IsExplicitlyNullInConfig(d *ResourceData, configFieldName string) bool {
	var isNull bool

	v, ok := d.GetRawConfig().AsValueMap()[configFieldName]

       if !ok {
         return false // As this is not defined in the schema
       }

	if v.IsNull() {
		isNull = true
	}

	return isNull
}
```

This then aligns better to the fw as the methods in fw only apply to properties defined in the schema.

Reply:
TBH if we add a state migration we remove the need for the call to `d.GetRawState()` entirely, since it'll then have a default value assigned - which it sounds like was missing from the 3.0 upgrade - so could we update this to use a state migration instead?

---

For the method above, that change makes sense to me, it's only used in a handful of places anyway if memory serves.

FWIW @mbfrahry is currently working on schema/model validation for Typed Data Sources/Resources which'll allow catching bugs where a field is provided to Get/Set but doesn't exist at unit test time, rather than acctest time - but that doesn't help for untyped resources (although we'll be migrating towards Typed Resources as time allows).

Reply:
@tombuildsstuff You are right, a state migration suffices here. I've made a new commit, please take another look.